### PR TITLE
[SR-13088] Fixing remaining spurious cast to unrelated type warning involving collection elements

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
This is a followup for resolution of SR-13088
This fixes SR-7187 and SR-6192 and possibly SR-7738
Also, this handles the unrelated warning for checked cast from protocol 
composition to unrelated non-existential. 

Resolves remaining issues on SR-13088.

### CODE REVIEW

File: lib/Sema/TypeCheckConstraints.cpp [hunk 1]
```diff
@@ -3264,7 +3267,9 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
       return castKind;
 
     case CheckedCastKind::Unresolved:
-      return failed();
+      if (contextKind == CheckedCastContextKind::Coercion)
+        return failed();
+      return castKind;
```
Comment (hamishknight) [inline lib/Sema/TypeCheckConstraints.cpp:3272]: Could you add a comment to explain this logic? Something along the lines of:

```cxx
Even if we know the elements cannot be downcast, we cannot return failed() here as
it's possible for an empty Array, Set or Dictionary to be cast to any element type.
The one exception to this is when we're checking whether we can treat a coercion as
a checked cast – we don't want to tell the user to use as!, as it's probably the
wrong suggestion.
```

File: lib/Sema/TypeCheckConstraints.cpp [hunk 2]
```diff
@@ -3295,15 +3300,17 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
                                          BridgingCoercion);
         break;
 
+      case CheckedCastKind::Unresolved:
+        if (contextKind == CheckedCastContextKind::Coercion)
+          return failed();
+        LLVM_FALLTHROUGH;
```
Comment (hamishknight) [inline lib/Sema/TypeCheckConstraints.cpp:3306]: Could you also add a comment here? It can be something brief like "Handled the same as in checkElementCast; see comment there for rationale."

File: lib/Sema/TypeCheckConstraints.cpp [hunk 3]
```diff
@@ -3318,15 +3325,17 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
...
```
Comment (hamishknight) [inline lib/Sema/TypeCheckConstraints.cpp:3331]: And the same comment here :)

**Review by hamishknight — COMMENTED** [review]
Thanks! I have a couple nits, but this looks good to me!

File: lib/Sema/TypeCheckConstraints.cpp (hunk 1 above)
Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckConstraints.cpp:3272]: Sure! I think this explains everything.
Thanks, @hamishknight  :)

Comment (LucianoPAlmeida) [reply lib/Sema/TypeCheckConstraints.cpp:3278]: Fixed!

**Review by hamishknight — APPROVED** [review]

**Comment by varungandhi-apple** [comment]
I like how you've covered the different edge cases with tests. 😄

**Comment by LucianoPAlmeida** [comment]
> I like how you've covered the different edge cases with tests. 😄

Thanks @varungandhi-apple :)

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test

**Review by xedin — APPROVED** [review]

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test Windows platform

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test Windows platform

**Comment by LucianoPAlmeida** [comment]
@swift-ci Please smoke test Windows platform

**Comment by hamishknight** [comment]
@LucianoPAlmeida The failing test `IRGen/autolink_merge_from_clangimporter.swift` is unrelated to your change and is being fixed in https://github.com/apple/swift/pull/32751. As such, feel free to merge this :)

**Comment by LucianoPAlmeida** [comment]
> @LucianoPAlmeida The failing test `IRGen/autolink_merge_from_clangimporter.swift` is unrelated to your change and is being fixed in #32751. As such, feel free to merge this :)

Ahh, thanks @hamishknight :)
